Add signal handler for SIGUSR1 to close logfile

thingsconnected opened the pull request:
The logfile will be re-opened automatically as soon as something is
logged. This allows for log rotation, by first renaming existing log
file(s) and then performing:

    killall -USR1 unison

In case no log file has been opened (yet), the signal is ignored.

This feature is intended for cases where unison is running continuously
as a background (daemon) process, e.g.

    unison -repeat watch

gdt commented:
This doesn't change the documentation.   I'm not sure if that's a fair comment.

Is there  a norm (in the POSIX world, presumably) that programs behave this way on SIGUSR1, or is this custom behavior?

thingsconnected commented:
As far as I know you are free to use SIGUSRn in any way you like. E.g. nginx anb mongodb use it for logfile rotation (says google, page 1 of search results). Other projects use SIGHUP for this purpose, but (existing) users may expect unison to quit on SIGHUP, which is the default action, so I thought it would be best to use SIGUSR1.

The alternative is to perform a restart of unison (like e.g. Apache does when rotating logs), which I think is a bit clunky, especially when continuously monitoring a filesystem for changes.

tleedjarv commented on src/ubase/trace.ml:
```diff
@@ -15,6 +15,10 @@
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *)
 
+(* ignore SIGUSR1 by default, until the logfile is opened, where a signal
+   handler is installed *)
+open Sys;;
+signal sigusr1 Signal_ignore;;
```
Why do you need to ignore the signal at all? What happens if you don't?

thingsconnected replied:
The signal handler is only installed as soon as the logfile is opened. If you don't explicitly ignore the signal, it will call the default handler and terminate the program.

tleedjarv replied:
Ok, I see. But why don't you install the handler immediately? There is no dependency on the logfile being opened, as far as I can see.

tleedjarv commented on src/ubase/trace.ml:
```diff
@@ -121,7 +125,16 @@ let logfile =
+  ;;
+
 let rec getLogch() =
+  ignore(signal sigusr1 (Signal_handle closelog));
```
Why set the signal handler over and over again instead of setting it once (e.g. like you currently set Signal_ignore)?

tleedjarv replied:
Also, just so you know, you can use `Sys.set_signal`. Then you don't have to use `ignore` here.

thingsconnected replied:
That would be fine as well. Given my total lack of experience in OCaml, I just went with this solution, as it compiled and worked after spending a bit too much time on it.

I have revisited the code and changed it to install the signal handler simply once. The commit has been pushed onto this pull request.

tleedjarv commented on src/ubase/trace.ml:
```diff
@@ -15,6 +15,10 @@
 
+(* ignore SIGUSR1 by default, until the logfile is opened, where a signal
+   handler is installed *)
+open Sys;;
```
Don't open `Sys`, use the `Sys.` notation instead.

tleedjarv commented on src/ubase/trace.ml:
```diff
@@ -121,7 +125,16 @@ let logfile =
+    None -> ()
+  | Some(ch,file) ->
+      close_out ch;
+      logch := None;
+  ;;
+
```
Just to illustrate what I had in mind.
```suggestion
      logch := None

let sigusr1_behavior = signal sigusr1 (Signal_handle closelog)
```
And then of course remove the `Signal_ignore` and installing the handler in function `getLogch`.

thingsconnected replied:
Before reading this I was already busy patching and testing, please check if my latest tweaks match your expectations.

tleedjarv replied:
No worries. Looks good to me.

thingsconnected commented:
> This doesn't change the documentation. I'm not sure if that's a fair comment.

I added explanation in (I hope) all the right places.

gdt commented:
Needs rebasing and then I think there is agreement it's ok.

Whether it goes in 2.51.3 or not is a different manner.  My own bias is that 2.51.3 is so overdue that only outright bugs with clear fixes should be fixed (and this doesn't qualify), and otherwise we should be heading to RC.   I would like to then see a 2.51.4 with as many non-breaking improvements merged, before making the breaking changes.

thingsconnected commented:
> Needs rebasing and then I think there is agreement it's ok.

I rebased it onto unison/master and squashed the two overlapping commits.

> Whether it goes in 2.51.3 or not is a different manner. My own bias is that 2.51.3 is so overdue that only outright bugs with clear fixes should be fixed (and this doesn't qualify), and otherwise we should be heading to RC. I would like to then see a 2.51.4 with as many non-breaking improvements merged, before making the breaking changes.

That's fine by me. It would be great to have this feature merged upstream some day.

bcpierce00 commented:
One might also consider maintaining both 2.51 and 2.52 branches, at least for a while…


> On Oct 12, 2020, at 4:28 PM, thingsconnected <notifications@github.com> wrote:
> 
> 
> Needs rebasing and then I think there is agreement it's ok.
> 
> I rebased it onto unison/master and squashed the two overlapping commits.
> 
> Whether it goes in 2.51.3 or not is a different manner. My own bias is that 2.51.3 is so overdue that only outright bugs with clear fixes should be fixed (and this doesn't qualify), and otherwise we should be heading to RC. I would like to then see a 2.51.4 with as many non-breaking improvements merged, before making the breaking changes.
> 
> That's fine by me. It would be great to have this feature merged upstream some day.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/bcpierce00/unison/pull/365#issuecomment-707329586>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABVQQC2PPIB5HKZDAQEVJQ3SKNRFBANCNFSM4QID4TSQ>.
>

gdt commented:
I propose to hit merge after the release.  Speak up for or against by Sunday, please.
